src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/SystemService.java

```diff
@@ -184,6 +190,82 @@ public AnchorFile getAnchorFile() throws AnchorNotFoundException {
         return anchorFile;
     }
 
+    /**
+     * Calculate the hex hash of the given anchor file. Used to verify/preview an anchor file before
+     * uploading it
+     * @param anchorBytes
+     * @return
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     */
+    public AnchorFile getAnchorFileFromBytes(byte[] anchorBytes) throws InvalidAnchorInstanceException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        AnchorFile anchorFile = new AnchorFile(calculateAnchorHexHash(anchorBytes));
+        anchorFile.setCreatedAt(FormatUtils.fromDateToOffsetDateTime(anchor.getGeneratedAt()));
+        return anchorFile;
+    }
+
+    /**
+     * Upload a new configuration anchor. A temporary anchor file is created on the filesystem in order to run
+     * the verification process with configuration-client module (via external script).
+     * @param anchorBytes
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     * @throws AnchorUploadException in case of external process exceptions
+     */
+    public void uploadAnchor(byte[] anchorBytes) throws InvalidAnchorInstanceException, AnchorUploadException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        File tempAnchor = null;
+        try {
+            tempAnchor = createTemporaryAnchorFile(anchorBytes);
+            configurationVerifier.verifyInternalConfiguration(tempAnchor.getAbsolutePath());
+            anchorRepository.saveAndReplace(tempAnchor);
```

Comment:
Nice catch! TBH this seems pretty important step 😄

Reply:
It looks like restTemplate.getForEntity has infinite timeout, so in a bad configuration it could hang forever. Or do you know how that is? Maybe something like 60s would be better.

I am not sure how old Ruby implementation does the same. It might not be any better.

Reply:
It really does have an infinite timeout by default -> not good. I added a 60s timeout for this request. Also the user gets an error code now if the request itself fails for some reason.